### DISCUSSION TITLE
Locking "track revisions" in generated DOCX

### DISCUSSION THREAD

**Discussion by osdiab** [dolanmiu/docx#1339] [Q&A]
Is there a way to restrict editing/only allow comments or force tracking changes, enforced with a password in a generated docx document? Like the `Review > Protect Document` button in Word? We want to generate such a Word document are a looking for the right library to do so.

[It is natively supported in Word](https://support.microsoft.com/en-us/office/make-a-document-read-only-5c25909c-46d9-4eb0-9d1f-d072a560e340), and it appears that [the Open XML SDK has a binding for this](https://docs.microsoft.com/en-us/dotnet/api/documentformat.openxml.wordprocessing.documentprotection?view=openxml-2.8.1), though idk how it works and don't really want to need to set up separate infrastructure to run this on our web service.

EDIT: just tried unzipping a docx file, saw the entry inside the `word/settings.xml` file - seems like some pretty simple XML to generate. Maybe I can contribute this if this type of functionality doesn't exist yet? Basically just this, where the crypto algorithm, password, and salt need to be provided - not sure how to use the `rsaAES` crypto provider Word generated by default, or how exactly it combines with the salt, but [these are the possible values](https://docs.microsoft.com/en-us/dotnet/api/documentformat.openxml.wordprocessing.cryptprovidervalues?view=openxml-2.8.1).

```xml
<w:documentProtection 
  w:edit="comments" 
  w:enforcement="1"
  w:cryptProviderType="rsaAES"
  w:cryptAlgorithmClass="hash"
  w:cryptAlgorithmType="typeAny"
  w:cryptAlgorithmSid="14"
  w:cryptSpinCount="100000"
  w:hash="mx/mgUHZgtV4x7qWe8+O2bfFuH4FkFQAURI2RGw+T9zQ0823foS7/01MWV4o/nfMhIVA+lSacCFOGVaanHqJsQ=="
  w:salt="9JSdHTo/rIs2X8nlixguTg=="
/>
```

Thanks!

**Comment by dolanmiu** [comment]
If you could create a PR for this, it would be great!

Found some more resources:

http://www.datypic.com/sc/ooxml/a-w_hash-1.html
https://c-rex.net/projects/samples/ooxml/e1/Part4/OOXML_P4_DOCX_writeProtection_topic_ID0EV4LY.html

I believe for the rsa AES stuff, we can use a library?

https://www.npmjs.com/package/hybrid-crypto-js

**Reply by osdiab** [reply]
Yeah just wasn't sure there was anything special about the Microsoft version of that, probably should be fine! I'll give it a go probably sometime early next year :)

**Reply by dolanmiu** [reply]
Nice one, looking forward!
